### PR TITLE
o/snapstate/check_snap.go: add support for many subversions in assumes snapdX..

### PR DESCRIPTION
**Pull request by anonymouse64**
This came up recently when we ended up doing a sub version release of snapd
2.48.2.1, and we wanted snaps to depend on this, but we couldn't make them do it
because the parsing regex gode gave up after 2.48.2. So make it fully support
any number of subversions in the hope that we never have to use this.

If folks think this will never become a problem, I call to the stand runC's release 
page: https://github.com/opencontainers/runc/releases . I hope and pray we never
get there, but who knows, it is better to have more support and never use it than 
need the support and not have it.

### CODE REVIEW

File: overlord/snapstate/check_snap.go [hunk 1]
```diff
@@ -196,15 +196,17 @@ func checkAssumes(si *snap.Info) error {
 	return nil
 }
 
-var versionExp = regexp.MustCompile(`^([1-9][0-9]*)(?:\.([0-9]+)(?:\.([0-9]+))?)?`)
+var versionExp = regexp.MustCompile(`^(?:[1-9][0-9]*)(?:\.(?:[0-9]+))*`)
```
Comment (bboozzoo) [inline overlord/snapstate/check_snap.go:199]:
```
// regular expression which matches a version expressed as groups of digits separated with dots
```

File: overlord/snapstate/check_snap_test.go [hunk 2]
```diff
@@ -136,12 +136,36 @@ var assumesTests = []struct {
...
+}, {
+	// the horror the horror!
+	assumes: "[snapd2.15.1.2.4.5.6.7.8.8]",
+	version: "2.15.1.2.4.5.6.7.8.8",
+}, {
+	assumes: "[snapd2.15.1.2.4.5.6.7.8.8]",
+	version: "2.15.1.2.4.5.6.7.8.9",
+}, {
+	assumes: "[snapd2.15.1.2]",
+	version: "2.15.1.3",
```
Comment (bboozzoo) [inline overlord/snapstate/check_snap_test.go:154]: If I'm reading this right, we could add a test case:
```
}, {
	assumes: "[snapd2]",
	version: "2.1",
}, {
```

**Review by bboozzoo — APPROVED** [review]

File: overlord/snapstate/check_snap_test.go (hunk 2 above)
Comment (bboozzoo) [reply overlord/snapstate/check_snap_test.go:154]: Nvm, it's just down below.

Comment (bboozzoo) [reply overlord/snapstate/check_snap_test.go:154]: Actually, we don't have:
```
	assumes: "[snapd2]",
	version: "3",
```
and
```
	assumes: "[snapd2]",
	version: "2",
```

Comment (anonymouse64) [reply overlord/snapstate/check_snap_test.go:154]: I added these test cases along with some others that we weren't covering in unit tests as well as the untested cases that the spread tests caught actually

File: overlord/snapstate/check_snap.go (hunk 1 above)
Comment (anonymouse64) [reply overlord/snapstate/check_snap.go:201]: expanded on this a bit and added it

File: overlord/snapstate/check_snap_test.go [hunk 3]
```diff
@@ -136,12 +140,46 @@ var assumesTests = []struct {
...
 	assumes: "[snapd2.15.1]",
 	version: "2.15.1",
+}, {
+	assumes: "[snapd2.15.1.2]",
+	version: "2.15.1.2",
+}, {
+	assumes: "[snapd2.15.1.2]",
+	version: "2.15.1.3",
+}, {
+	// the horror the horror!
```
Comment (mvo5) [inline overlord/snapstate/check_snap_test.go:150]: hahaha

**Review by mvo5 — APPROVED** [review]
Thanks for fixing this

**Pull request merged**